### DISCUSSION TITLE
Performance question

### DISCUSSION THREAD

**Discussion by matrem** [atteneder/glTFast#468] [General]
Hi,

I have some questions regarding performances :
- When the documentation says : 
   "tries not to block the main thread for longer than a certain time budget"
It's only true for loading not instantiation ?
So when calling InstantiateMainScene, we have a risk of render thread lag if object is to heavy?
- C# managed memory is used, so garbage collector can be called at any moment causing all threads to lag ?
I experienced a big 3 sec unity freeze at gltf.load for a big model with many textures (for a total of 300MB), and the was during memory allocation/collection.

**Comment by atteneder** [comment]
> Hi,
> 
> I have some questions regarding performances :
> 
>     * When the documentation says :
>       "tries not to block the main thread for longer than a certain time budget"
>       It's only true for loading not instantiation ?
>       So when calling InstantiateMainScene, we have a risk of render thread lag if object is to heavy?

That's true for the current version 4.x and before. In the `main` branch (or the upcoming 5.x version) instantiation is also async!

>     * C# managed memory is used, so garbage collector can be called at any moment causing all threads to lag ?
>       I experienced a big 3 sec unity freeze at gltf.load for a big model with many textures (for a total of 300MB), and the was during memory allocation/collection.

I'm currently investigating and optimizing in this area wrt KTX textures and I'm on to something. What texture formats do you use and on which platform/hardware? Feel free to provide demo content. That always helps get a better understanding of user's needs.

**Reply by atteneder** [reply]
I've also documented the async instantiation progress here:

<https://pixel.engineer/posts/gltfast-async-instantiation/>

expect some improvement on complex scenes!
